Makefile

```diff
@@ -1,31 +1,52 @@
+# Unless set otherwise, the container runtime is Docker
+DOCKER ?= docker
+
+prog ?= configurator
+git_tag := $(shell git describe --abbrev=0 --tags)
+sources := main.go $(wildcard cmd/*.go)
+plugin_source_prefix := pkg/generator/plugins
+plugin_sources := $(filter-out %_test.go,$(wildcard $(plugin_source_prefix)/*/*.go))
+plugin_binaries := $(addprefix lib/,$(patsubst %.go,%.so,$(notdir $(plugin_sources))))
 
 # build everything at once
+.PHONY: all
 all: plugins exe test
 
 # build the main executable to make configs
+.PHONY: main driver binaries exe
 main: exe
 driver: exe
-exe:
-	go build --tags=all -o configurator
+binaries: exe
+exe: $(prog)
+
+# build named executable from go sources
+$(prog): $(sources)
+	go build --tags=all -o $(prog)
+
+.PHONY: container
+container: binaries plugins
+	$(DOCKER) build . --build-arg --no-cache --pull --tag '$(prog):$(git_tag)-dirty'
+
+.PHONY: container-testing
+container-testing: binaries plugins
+	$(DOCKER) build . --tag $(prog):testing
 
 # build all of the generators into plugins
-plugins:
+.PHONY: plugins
+plugins: $(plugin_binaries)
+
+# how to make each plugin
+lib/%.so: pkg/generator/plugins/%/*.go
 	mkdir -p lib
-	go build -buildmode=plugin -o lib/conman.so internal/generator/plugins/conman/conman.go
-	go build -buildmode=plugin -o lib/coredhcp.so internal/generator/plugins/coredhcp/coredhcp.go
-	go build -buildmode=plugin -o lib/dhcpd.so internal/generator/plugins/dhcpd/dhcpd.go
-	go build -buildmode=plugin -o lib/dnsmasq.so internal/generator/plugins/dnsmasq/dnsmasq.go
-	go build -buildmode=plugin -o lib/example.so internal/generator/plugins/example/example.go
-	go build -buildmode=plugin -o lib/hostfile.so internal/generator/plugins/hostfile/hostfile.go
-	go build -buildmode=plugin -o lib/powerman.so internal/generator/plugins/powerman/powerman.go
-	go build -buildmode=plugin -o lib/syslog.so internal/generator/plugins/syslog/syslog.go
-	go build -buildmode=plugin -o lib/warewulf.so internal/generator/plugins/warewulf/warewulf.go
+	go build -buildmode=plugin -o $@ $<
 
 # remove executable and all built plugins
+.PHONY: clean
 clean:
-	rm configurator
-	rm lib/*
+	rm -f configurator
+	rm -f lib/*
 
 # run all of the unit tests
+.PHONY: test
 test:
```

Comment:
I think it would be a good idea to have the plugins and main program binary by prerequisites of this target. That way, running `make test` on a fresh checkout would build first before testing. E.g:

```make
test: $(prog) $(plugin_binaries)
```

Reply:
I might change this so the main program is also being build in the testing directory. That way, we should be able to ensure that the main driver is built at the same time as the plugins.